### PR TITLE
Preview: DNP3 Support - v8

### PR DESCRIPTION
**Pull request by jasonish**
Replaces: #1844 
Remine ticket: https://redmine.openinfosecfoundation.org/issues/1745

This PR contains:
An application decoder for DNP3.
* Snort compatible DNP3 keywords:
  * dnp3_data
  * dnp3_func
  * dnp3_ind
  * dnp3_obj
* A DNP3 output module for EVE (JSON logging).
* DNP3 support for Lua detection.

Documentation for the DNP3 rule keywords can be found here:
https://redmine.openinfosecfoundation.org/projects/suricata/wiki/DNP3_keyword
OR in newer format:
http://jasonish-suricata.readthedocs.org/en/latest/rules/dnp3-keywords.html

Changes from the last PR include:
* Fixups from last PR comments.
* Many more objects supported.
* Fixups found during fuzzing.
* Other cleanups.

PRScript output:
- PR jasonish-pcap: https://buildbot.openinfosecfoundation.org/builders/jasonish-pcap/builds/215
- PR jasonish: https://buildbot.openinfosecfoundation.org/builders/jasonish/builds/220

### CODE REVIEW

File: scripts/dnp3-gen/dnp3-gen.py [hunk 1]
```diff
@@ -0,0 +1,707 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * DO NOT EDIT. THIS FILE IS AUTO-GENERATED.
```
Comment (inliniac) [inline scripts/dnp3-gen/dnp3-gen.py]: can this include the exact command used to generate the file?

File: src/detect.h [hunk 2]
```diff
@@ -127,6 +127,10 @@ enum DetectSigmatchListEnum {
 
     DETECT_SM_LIST_BASE64_DATA,
 
+    DETECT_SM_LIST_DNP3_DATA_MATCH,
+    DETECT_SM_LIST_DNP3_MATCH,
```
Comment (inliniac) [inline src/detect.h]: can DETECT_SM_LIST_DNP3_MATCH and DETECT_SM_LIST_DNP3_LUA_MATCH be merged into a single list?

File: src/app-layer-dnp3-decoders.c [hunk 3]
```diff
@@ -0,0 +1,1930 @@
+
```
Comment (inliniac) [inline src/app-layer-dnp3-decoders.c]: missing copyright header

**Comment by inliniac** [comment]
Part of the code is generated by a script, but it's not done like that by the build system correct? Is that how it should be? When & how often would we need to run this script?

**Comment by jasonish** [comment]
> Part of the code is generated by a script, but it's not done like that by the build system correct? Is that how it should be? When & how often would we need to run this script?

You would need to regenerate after editing scripts/dnp3-gen/dnp3-objects.yaml which you would edit when adding a new object, or fixing an old one.

You would also need to regenerate if fixing the code in one of the generated code sections, say refactoring.

I thought about doing it at part of the build, but I didn't based on some code gen I have worked with in the past with BSD kernels where the generation is done once, committed.  Then only the developer tweaking the definitions at a later date need worry about regenerating.  I think each method has its pros and cons, but I like the code to be readable after a checkout before build.

File: src/app-layer-dnp3-decoders.c (hunk 3 above)
Comment (jasonish) [inline src/app-layer-dnp3-decoders.c]: Oops, that file shouldn't be there.

File: scripts/dnp3-gen/dnp3-gen.py (hunk 1 above)
Comment (jasonish) [inline scripts/dnp3-gen/dnp3-gen.py]: Ok, all generated files, or segments have a comment, right now it looks like:

/* Code generated by:
 *     ./scripts/dnp3-gen/dnp3-gen.py
 */

File: src/detect.h (hunk 2 above)
Comment (jasonish) [inline src/detect.h]: Yes, with a bit of reworking it looks possible.

Comment (jasonish) [inline src/detect.h]: Looks like this would involve converting the detection functions to AppLayerTxMatch variants. Is that what you had in mind?

Comment (jasonish) [inline src/detect.h]: See https://github.com/inliniac/suricata/pull/2104/commits/dba95ad3e2a91fdd0bb068a5ff988d7100fa2184

**Pull request closed without merging**